# Conexión MySQL con VSC

In [41]:
import pymysql
import pandas as pd

try:
    # Conexión a la base de datos del simulador
    connection = pymysql.connect(
        host='212.227.90.6',
        user='Equipo12',
        password='E1q2u3i4p5o12',
        database='Equip_12',
        port=3306
    )
    print("✅ Conexión establecida con MySQL Server")

    # Crear un cursor
    cursor = connection.cursor()

    # Consultar las tablas disponibles
    cursor.execute("SHOW TABLES;")
    tables = cursor.fetchall()

    # Diccionario para almacenar los dataframes
    dataframes = {}

    for tabla in tables:
        table_name = tabla[0]
        print(f"📥 Cargando tabla: {table_name}")
        query = f"SELECT * FROM {table_name};"
        df = pd.read_sql(query, connection)
        dataframes[table_name] = df
        print(f"✅ DataFrame para '{table_name}' creado con {df.shape[0]} filas")

except pymysql.MySQLError as e:
    print("❌ Error al conectar con MySQL:", e)


✅ Conexión establecida con MySQL Server
📥 Cargando tabla: RRHH
✅ DataFrame para 'RRHH' creado con 740 filas


C:\Users\dessb\AppData\Local\Temp\ipykernel_6160\2277306723.py:29: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


In [42]:
df_rrhh = dataframes['RRHH']

In [43]:
df_rrhh.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 740 entries, 0 to 739
Data columns (total 21 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   ID                       740 non-null    int64 
 1   Reason_absence           740 non-null    int64 
 2   Month_absence            740 non-null    int64 
 3   Day_week                 740 non-null    int64 
 4   Seasons                  740 non-null    int64 
 5   Transportation_expense   740 non-null    int64 
 6   Distance_Residence_Work  740 non-null    int64 
 7   Service_time             740 non-null    int64 
 8   Age                      740 non-null    int64 
 9   Work_load_Average_day    740 non-null    object
 10  Hit_target               740 non-null    int64 
 11  Disciplinary_failure     740 non-null    object
 12  Education                740 non-null    object
 13  Son                      740 non-null    object
 14  Social_drinker           740 non-null    o

# Data Cleaning
## Tratamiento de valores faltantes

No hay valores nulos, por lo tanto no se hace ninguna modificación

In [44]:
df_rrhh.isna().sum()

ID                         0
Reason_absence             0
Month_absence              0
Day_week                   0
Seasons                    0
Transportation_expense     0
Distance_Residence_Work    0
Service_time               0
Age                        0
Work_load_Average_day      0
Hit_target                 0
Disciplinary_failure       0
Education                  0
Son                        0
Social_drinker             0
Social_smoker              0
Pet                        0
Weight                     0
Height                     0
Body_mass_index            0
Absenteeism_hours          0
dtype: int64

# Corrección de tipos de datos

Antes de eliminar duplicados, se recomienda la corrección de tipos de datos:

- Reemplazar comas por puntos (Work_load_Average_day) debería ser float (número)
- Disciplinary_failure, Social_drinker, Social_smoker, Education, Son y Pet aunque sus valores son numéricos y deberían ser enteros (int) o categorías (category), están codificados como texto.


In [45]:
df_rrhh.dtypes

ID                          int64
Reason_absence              int64
Month_absence               int64
Day_week                    int64
Seasons                     int64
Transportation_expense      int64
Distance_Residence_Work     int64
Service_time                int64
Age                         int64
Work_load_Average_day      object
Hit_target                  int64
Disciplinary_failure       object
Education                  object
Son                        object
Social_drinker             object
Social_smoker              object
Pet                        object
Weight                      int64
Height                      int64
Body_mass_index             int64
Absenteeism_hours           int64
dtype: object

In [46]:
df_rrhh["Work_load_Average_day"] = df_rrhh["Work_load_Average_day"].str.replace(",", ".").astype(float)

In [47]:
columnas_a_convertir = ["Disciplinary_failure", "Education", "Son", "Social_drinker", "Social_smoker", "Pet"]
df_rrhh[columnas_a_convertir] = df_rrhh[columnas_a_convertir].astype(int)

In [48]:
df_rrhh.dtypes

ID                           int64
Reason_absence               int64
Month_absence                int64
Day_week                     int64
Seasons                      int64
Transportation_expense       int64
Distance_Residence_Work      int64
Service_time                 int64
Age                          int64
Work_load_Average_day      float64
Hit_target                   int64
Disciplinary_failure         int64
Education                    int64
Son                          int64
Social_drinker               int64
Social_smoker                int64
Pet                          int64
Weight                       int64
Height                       int64
Body_mass_index              int64
Absenteeism_hours            int64
dtype: object

# Eliminación o corrección de duplicados

In [49]:
# Duplicados completos
df_rrhh.duplicated().sum()

np.int64(34)

In [50]:
df_rrhh[df_rrhh.duplicated()].sort_values(by='ID')

# pd.set_option('display.max_columns', None) 
# # para ver todas las columnas y que no se corten por el limite de Panda

ID  Reason_absence  Month_absence  Day_week  Seasons  \
386   3              27              2         4        2   
383   3              27              2         4        2   
388   3              27              2         4        2   
387   3              27              2         6        2   
389   3              27              2         6        2   
454   3              23              7         6        1   
434   3              27              3         5        2   
431   3              27              3         5        2   
563   3              27              2         6        2   
582   3              27              3         4        2   
577   3              27              3         4        2   
570   3              27              2         5        2   
569   3              27              2         4        2   
564   3              27              2         2        2   
562   3              27              2         4        2   
663   3              28             12         6        4   
501   5              23             10         2        4   
92   10              22             12         4        4   
599  15              28              5         5        3   
590  22              27              4         6        3   
600  22              27              5         6        3   
589  22              27              4         6        3   
396  22              23              5         4        3   
592  22              27              4         6        3   
679  24              28              9         3        1   
528  24              28              7         3        1   
482  27              23              1         5        2   
412  28              23             12         4        4   
617  28              23             11         4        4   
480  34              27              1         3        2   
481  34              27              1         4        2   
348  34              23             10         3        4   
483  34              27              1         5        2   
479  34              27              1         2        2   

     Transportation_expense  Distance_Residence_Work  Service_time  Age  \
386                     179                       51            18   38   
383                     179                       51            18   38   
388                     179                       51            18   38   
387                     179                       51            18   38   
389                     179                       51            18   38   
454                     179                       51            18   38   
434                     179                       51            18   38   
431                     179                       51            18   38   
563                     179                       51            18   38   
582                     179                       51            18   38   
577                     179                       51            18   38   
570                     179                       51            18   38   
569                     179                       51            18   38   
564                     179                       51            18   38   
562                     179                       51            18   38   
663                     179                       51            18   38   
501                     235                       20            13   43   
92                      361                       52             3   28   
599                     291                       31            12   40   
590                     179                       26             9   30   
600                     179                       26             9   30   
589                     179                       26             9   30   
396                     179                       26             9   30   
592                     179                       26             9   

In [51]:
# Lista de filas duplicadas exactas a eliminar
filas_a_eliminar = [383, 388, 389, 431, 577, 562, 589, 592]

# Eliminar esas filas del DataFrame
df_rrhh = df_rrhh.drop(index=filas_a_eliminar).reset_index(drop=True)

In [52]:
# comprobación de que ya no están los duplicados:
df_rrhh[df_rrhh.duplicated()].sort_values(by='ID')

ID  Reason_absence  Month_absence  Day_week  Seasons  \
386   3              27              2         6        2   
385   3              27              2         4        2   
430   3              27              3         5        2   
450   3              23              7         6        1   
558   3              27              2         6        2   
565   3              27              2         5        2   
564   3              27              2         4        2   
559   3              27              2         2        2   
576   3              27              3         4        2   
655   3              28             12         6        4   
497   5              23             10         2        4   
92   10              22             12         4        4   
591  15              28              5         5        3   
592  22              27              5         6        3   
583  22              27              4         6        3   
393  22              23              5         4        3   
671  24              28              9         3        1   
524  24              28              7         3        1   
478  27              23              1         5        2   
609  28              23             11         4        4   
409  28              23             12         4        4   
348  34              23             10         3        4   
476  34              27              1         3        2   
477  34              27              1         4        2   
479  34              27              1         5        2   
475  34              27              1         2        2   

     Transportation_expense  Distance_Residence_Work  Service_time  Age  \
386                     179                       51            18   38   
385                     179                       51            18   38   
430                     179                       51            18   38   
450                     179                       51            18   38   
558                     179                       51            18   38   
565                     179                       51            18   38   
564                     179                       51            18   38   
559                     179                       51            18   38   
576                     179                       51            18   38   
655                     179                       51            18   38   
497                     235                       20            13   43   
92                      361                       52             3   28   
591                     291                       31            12   40   
592                     179                       26             9   30   
583                     179                       26             9   30   
393                     179                       26             9   30   
671                     246                       25            16   41   
524                     246                       25            16   41   
478                     184                       42             7   27   
609                     225                       26             9   28   
409                     225                       26             9   28   
348                     118                       10            10   37   
476                     118                       10            10   37   
477                     118                       10            10   37   
479                     118                       10            10   37   
475                     118                       10            10   37   

     Work_load_Average_day  Hit_target  Disciplinary_failure  Education  Son  \
386                251.818          96                     0          1    0   
385                251.818          96                     0          1    0   
430                222.196          99                     0          1    0   
450        

# Validación de valores atípicos (outliers)

In [53]:
df_rrhh.describe()

ID  Reason_absence  Month_absence    Day_week     Seasons  \
count  732.000000      732.000000     732.000000  732.000000  732.000000   
mean    18.129781       19.131148       6.363388    3.904372    2.547814   
std     10.995004        8.439809       3.433428    1.422559    1.116550   
min      1.000000        0.000000       0.000000    2.000000    1.000000   
25%     10.000000       13.000000       3.000000    3.000000    2.000000   
50%     18.000000       23.000000       6.000000    4.000000    3.000000   
75%     28.000000       26.000000       9.250000    5.000000    4.000000   
max     36.000000       28.000000      12.000000    6.000000    4.000000   

       Transportation_expense  Distance_Residence_Work  Service_time  \
count              732.000000               732.000000    732.000000   
mean               221.792350                29.465847     12.519126   
std                 67.170185                14.789455      4.377022   
min                118.000000                 5.000000      1.000000   
25%                179.000000                16.000000      9.000000   
50%                225.000000                26.000000     13.000000   
75%                260.000000                49.250000     16.000000   
max                388.000000                52.000000     29.000000   

              Age  Work_load_Average_day  Hit_target  Disciplinary_failure  \
count  732.000000             732.000000  732.000000            732.000000   
mean    36.454918             271.784294   94.576503              0.054645   
std      6.503867              39.142040    3.786128              0.227441   
min     27.000000             205.917000   81.000000              0.000000   
25%     31.000000             244.387000   93.000000              0.000000   
50%     37.000000             264.604000   95.000000              0.000000   
75%     40.000000             294.217000   97.000000              0.000000   
max     58.000000             378.884000  100.000000              1.000000   

        Education         Son  Social_drinker  Social_smoker         Pet  \
count  732.000000  732.000000      732.000000     732.000000  732.000000   
mean     1.289617    1.030055        0.565574       0.073770    0.754098   
std      0.670465    1.099271        0.496020       0.261576    1.323128   
min      1.000000    0.000000        0.000000       0.000000    0.000000   
25%      1.000000    0.000000        0.000000       0.000000    0.000000   
50%      1.000000    1.000000        1.000000       0.000000    0.000000   
75%      1.000000    2.000000        1.000000       0.000000    1.000000   
max      4.000000    4.000000        1.000000       1.000000    8.000000   

           Weight      Height  Body_mass_index  Absenteeism_hours  
count  732.000000  732.000000       732.000000         732.000000  
mean    79.016393  172.135246        26.662568           6.972678  
std     12.865707    6.064587         4.272147          13.395563  
min     56.000000  163.000000        19.000000           0.000000  
25%     69.000000  169.000000        24.000000           2.000000  
50%     80.000000  170.000000        25.000000           3.000000  
75%     89.000000  172.000000        31.000000           8.000000  
max    108.000000  196.000000        38.000000         120.000000

Se han encontrado posibles valores extremos en:
- Transportation_expense (máx. 388), la media es 221. Esto sugiere que algunos empleados tienen gastos de transporte muy por encima del promedio. Podrían ser outliers, o casos especiales (viven lejos, taxis, etc.).
- Absenteeism_hours (máx. 120)
- Work_load_Average_day (máx. 310), muy cercano al máximo permitido en carga diaria

# Estandarización de formatos

In [54]:
df_rrhh['Month_absence'].value_counts()

Month_absence
3     85
10    71
2     68
7     67
5     64
11    63
6     54
8     54
9     53
4     51
1     50
12    49
0      3
Name: count, dtype: int64

En la columna Month_absence se ha detectado el valor 0, que no corresponde a ningún mes válido (1–12). Solo hay 3 registros con este valor. Clasificado como " mes desconocido"


In [55]:
# Convertir la columna a tipo string para permitir texto
df_rrhh['Month_absence'] = df_rrhh['Month_absence'].astype(str)

# Reemplazar '0' por 'mes desconocido'
df_rrhh['Month_absence'] = df_rrhh['Month_absence'].replace('0', 'mes desconocido')


In [56]:
df_rrhh['Month_absence'].value_counts()


Month_absence
3                  85
10                 71
2                  68
7                  67
5                  64
11                 63
6                  54
8                  54
9                  53
4                  51
1                  50
12                 49
mes desconocido     3
Name: count, dtype: int64

In [57]:
df_rrhh['Disciplinary_failure'].value_counts()
# está ok

Disciplinary_failure
0    692
1     40
Name: count, dtype: int64

In [58]:
df_rrhh['Education'].value_counts()
# está ok

Education
1    605
3     77
2     46
4      4
Name: count, dtype: int64

In [59]:
df_rrhh['Social_drinker'].value_counts()
# está ok

Social_drinker
1    414
0    318
Name: count, dtype: int64

In [60]:
df_rrhh['Social_smoker'].value_counts()
# está ok

Social_smoker
0    678
1     54
Name: count, dtype: int64

# Revisión de la variable edad

In [61]:
# Encontrar la edad máxima por empleado
max_ages = df_rrhh.groupby('ID')['Age'].max()


In [62]:
# Crear y anexar nueva columna llamada 'max_age' que contiene la mayor edad reportada de c/empleado

df_rrhh['Max_Age'] = df_rrhh['ID'].map(max_ages)

In [63]:
# Ver los valores que son diferentes entre 'age' y 'max_age'

df_rrhh[df_rrhh['Age'] != df_rrhh['Max_Age']][['ID', 'Age', 'Max_Age']]


ID  Age  Max_Age
690  29   28       41

In [64]:
# Busco los registros del empleado 29

df_rrhh[df_rrhh['ID'] == 29][['ID', 'Age']]

ID  Age
252  29   41
253  29   41
437  29   41
551  29   41
690  29   28

Únicamente hay un discrepancia de edad en el registro 29. Los demás son iguales.

In [65]:
# Inspección visual para ver que sí son iguales los registros de edad en los demás empleados 

df_rrhh[['ID', 'Age', 'Max_Age']].sort_values(by=['ID', 'Age'])


ID  Age  Max_Age
50    1   37       37
64    1   37       37
89    1   37       37
145   1   37       37
178   1   37       37
..   ..  ...      ...
692  36   50       50
694  36   50       50
706  36   50       50
719  36   50       50
725  36   50       50

[732 rows x 3 columns]

# Cambio del ID 29 dispar por ID 37

In [66]:
df_rrhh['ID'].nunique()

36

In [67]:
df_rrhh['ID'].sort_values().unique()


array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36])

In [68]:
#Sustitución del ID
df_rrhh.loc[(df_rrhh['ID'] == 29) & (df_rrhh['Age'] == 28), 'ID'] = 37

In [69]:
# Comprobación 

df_rrhh[df_rrhh['ID'] == 37]

ID  Reason_absence Month_absence  Day_week  Seasons  \
690  37               0             9         2        4   

     Transportation_expense  Distance_Residence_Work  Service_time  Age  \
690                     225                       26             9   28   

     Work_load_Average_day  Hit_target  Disciplinary_failure  Education  Son  \
690                241.476          92                     1          1    1   

     Social_drinker  Social_smoker  Pet  Weight  Height  Body_mass_index  \
690               0              0    2      69     169               24   

     Absenteeism_hours  Max_Age  
690                  0       41

In [70]:
# Comprobación 

df_rrhh[(df_rrhh['ID'] == 29) & (df_rrhh['Age'] == 28)]

Empty DataFrame
Columns: [ID, Reason_absence, Month_absence, Day_week, Seasons, Transportation_expense, Distance_Residence_Work, Service_time, Age, Work_load_Average_day, Hit_target, Disciplinary_failure, Education, Son, Social_drinker, Social_smoker, Pet, Weight, Height, Body_mass_index, Absenteeism_hours, Max_Age]
Index: []

# Estimación de la duración de la ausencia

In [71]:
# Asumo que la jornada laboral es de 8 horas 
df_rrhh['Estimated_Absence_Days'] = (df_rrhh['Absenteeism_hours'] / 8).round(2)


In [72]:
# Inspección visual 
df_rrhh[['ID', 'Absenteeism_hours', 'Estimated_Absence_Days']].head()

ID  Absenteeism_hours  Estimated_Absence_Days
0  14                120                    15.0
1  36                120                    15.0
2   9                120                    15.0
3  28                112                    14.0
4   9                112                    14.0

In [73]:
# Inspección visual
df_rrhh[df_rrhh['ID'] == 1][['ID', 'Absenteeism_hours', 'Estimated_Absence_Days']]


ID  Absenteeism_hours  Estimated_Absence_Days
50    1                 16                    2.00
64    1                  8                    1.00
89    1                  8                    1.00
145   1                  8                    1.00
178   1                  8                    1.00
182   1                  8                    1.00
193   1                  8                    1.00
196   1                  8                    1.00
247   1                  8                    1.00
250   1                  8                    1.00
285   1                  4                    0.50
303   1                  4                    0.50
307   1                  4                    0.50
328   1                  4                    0.50
338   1                  4                    0.50
364   1                  3                    0.38
377   1                  3                    0.38
499   1                  2                    0.25
501   1                  2                    0.25
638   1                  1                    0.12
647   1                  1                    0.12
656   1                  1                    0.12
717   1                  0                    0.00

# Reason_absence: reemplazo del 0 por 'motivo desconocido'

In [76]:
df_rrhh.columns

Index(['ID', 'Reason_absence', 'Month_absence', 'Day_week', 'Seasons',
       'Transportation_expense', 'Distance_Residence_Work', 'Service_time',
       'Age', 'Work_load_Average_day', 'Hit_target', 'Disciplinary_failure',
       'Education', 'Son', 'Social_drinker', 'Social_smoker', 'Pet', 'Weight',
       'Height', 'Body_mass_index', 'Absenteeism_hours', 'Max_Age',
       'Estimated_Absence_Days'],
      dtype='object')

In [77]:
df_rrhh['Reason_absence'] = df_rrhh['Reason_absence'].replace(0, 'motivo desconocido')

In [78]:
df_rrhh['Reason_absence'].value_counts()


Reason_absence
23                    149
28                    112
27                     61
13                     55
motivo desconocido     43
19                     40
22                     38
26                     33
25                     31
11                     26
10                     25
18                     21
14                     19
1                      16
7                      15
6                       8
12                      8
8                       6
21                      6
9                       4
16                      3
5                       3
24                      3
15                      2
4                       2
2                       1
17                      1
3                       1
Name: count, dtype: int64

43 columnas están registradas como 'motivo desconocido'

# Exportación del Dataframe

In [ ]:
# Archivo csv

df_rrhh.to_csv('df_rrhh_final.csv', index=False)



# Se puede abrir con Excel, Google Sheets o pandas.
# No guarda los tipos de datos especiales.



In [ ]:
# Pickle - Python native 

df_rrhh.to_pickle('df_rrhh_final.pkl')

# Guarda el DataFrame tal cual (con tipos, fechas, etc.)
# Solo funciona con Python.


In [ ]:
# JSON 
df_rrhh.to_json('df_rrhh_final.json', orient='records', lines=True)

#  Puede usarse con pandas.
